### PR TITLE
Enable HTML formatted notifications outside of product (migration from v2)

### PR DESCRIPTION
**Pull request by shadeon**
This PR enables the ability to setup notification channels in SCOM with HTML formatting enabled, and optionally SMTP flags for importance set, without having to use Squared Up.

As such, this has been generalized to allow the SCOM web console URL to be injected as well, and no longer requires explicit cloning of an existing channel (you can create one using the script).

Included (though not specific to this script) is a Pester file for ensuring scripts meet an appropriate level of quality.  If you are on Win 10 you already have the pester module and PSScriptAnalyser modules installed, but i'd recommend updating both modules first prior to invoking the tests.

### CODE REVIEW

File: GenericPester.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+{
+    $scriptPath = $script.FullName
+    Describe "$scriptPath" {    
+        $astErrors = @()
+        # Parse the function using AST
+        $ast = [System.Management.Automation.Language.Parser]::ParseFile($scriptPath, [ref]$null, [ref]$astErrors)
+
+        Context 'Structure and Syntax' {
+            It 'Contains no errors' { $astErrors.Count | Should BeLessThan 1 }
+            It 'Has a valid signiture (if signed)' {
```
Comment (clarkd) [inline GenericPester.Tests.ps1]: "signature"

File: GenericPester.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                $result = Get-AuthenticodeSignature $scriptPath
+                if ($result.Status -eq [System.Management.Automation.SignatureStatus]::NotSigned) {
```
Comment (clarkd) [inline GenericPester.Tests.ps1:26]: Could add a test for "is signed" as all our scripts should be.

File: New-SCOMHtmlNotificationChannel.ps1 [hunk 3]
```diff
@@ -0,0 +1,500 @@
+<#
+.SYNOPSIS
+    Creates SCOM notification channels with HTML formatting enabled.
+.DESCRIPTION
+    This script will create SCOM notification channels with HTML formatting options enabled. This allows for significantly richer notification templates.
+
+    Channels can either be created from scratch or clone their settings from an existing channel (and will therefore use it's predefined endpoints).
```
Comment (clarkd) [inline New-SCOMHtmlNotificationChannel.ps1]: it's -> its

File: New-SCOMHtmlNotificationChannel.ps1 [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+    By default notification bodies will include appropriate links to the SCOM Web console, but by specifying the SquaredUpUrl parameter links will use Squared Up instead.
```
Comment (clarkd) [inline New-SCOMHtmlNotificationChannel.ps1]: Should be "web console" or "Web Console"

Comment (clarkd) [reply New-SCOMHtmlNotificationChannel.ps1]: "SquaredUpURL" as per parameter name, and perhaps prefix with a "-" to indicate it is a parameter.

File: New-SCOMHtmlNotificationChannel.ps1 [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+    Creates High Importance HTML channels in SCOM using the specified SMTP relay settings.
+.EXAMPLE
+    C:\PS> .\New-SCOMHtmlNotificationChannel.ps1 -SquaredUpURL "SQUP.contoso.com/SquaredUpv3" -BaseSmtpChannelGuid 2149e02e-6bb2-661a-1535-17d1dba162ab -PlainText
+    Creates Plaintext channels with normal importance, taking all SMTP settings from an existing SMTP notification channel.
+.INPUTS
+    None
+.OUTPUTS
+    Microsoft.EnterpriseManagement.Administration.SmtpNotificationAction
+.NOTES 
+    Copyright 2016 Squared Up Limited, All Rights Reserved.
```
Comment (clarkd) [inline New-SCOMHtmlNotificationChannel.ps1]: 2017

File: New-SCOMHtmlNotificationChannel.ps1 [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    Creates SCOM notification channels with HTML formatting enabled.
+.DESCRIPTION
+    This script will create SCOM notification channels with HTML formatting options enabled. This allows for significantly richer notification templates.
+
+    Channels can either be created from scratch or clone their settings from an existing channel (and will therefore use it's predefined endpoints).
+
+    By default notification bodies will include appropriate links to the SCOM Web console, but by specifying the SquaredUpUrl parameter links will use Squared Up instead.
+
+    Please ensure that a management group connection exists prior to running the script - either run this from the Operations Manager shell or run New-SCOMManagementGroupConnection.
+.PARAMETER SquaredUpURL
```
Comment (clarkd) [inline New-SCOMHtmlNotificationChannel.ps1]: Should indicate that there should be no trailing slash.

File: New-SCOMHtmlNotificationChannel.ps1 [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+    $ManagedEntityDisplayName = '$Data/Context/DataItem/ManagedEntityDisplayName$'
+    $ManagedEntityPath = '$Data/Context/DataItem/ManagedEntityPath$'
+    $NotificationSubId = '$MPElement$'
+
+    if ($Html) 
+    {
+        return [string]::Join("`n", @(
+            "<!DOCTYPE html>",
+            "<html><head>",
+            "<!-- (c) Squared Up Ltd 2016 -->",
```
Comment (clarkd) [inline New-SCOMHtmlNotificationChannel.ps1]: 2017

File: New-SCOMHtmlNotificationChannel.ps1 [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+function Convert-SquaredUpUrl {
+    [CmdletBinding()]
+    [OutputType([string])]
+    Param(
+        [string]$Url
+    )
+
+    $Url  = $Url.TrimEnd('/')
```
Comment (clarkd) [inline New-SCOMHtmlNotificationChannel.ps1]: Ah just seen this - ignore the above :)

**Review by clarkd — CHANGES_REQUESTED** [review]

File: New-SCOMHtmlNotificationChannel.ps1 (hunk 5 above)
Comment (shadeon) [reply New-SCOMHtmlNotificationChannel.ps1]: Oh god, it's 2017 already.  When did it become 2017.

File: GenericPester.Tests.ps1 (hunk 2 above)
Comment (shadeon) [reply GenericPester.Tests.ps1:26]: Hmmm, from download.squaredup.com sure but they shouldn't be signed in-repo (contributors shouldn't be required to have a trusted code signing cert, never mind ours).

**Review by clarkd — APPROVED** [review]

**Comment by shadeon** [comment]
@richbenwell Any thoughts on this?  Just writing up the documentation so want to know if we're good to go or further changes are required.

**Review by richbenwell — APPROVED** [review]

**Pull request merged**